README.md
XBMC Addon Manager
==================

[![Build Status](https://secure.travis-ci.org/jbeluch/xam.png)](http://travis-ci.org/jbeluch/xam)

A CLI utility for searching and listing XBMC Addon Repositories.

Features
--------
* List addons available in remote repositories.
* List addons which require another addon as a dependency.
* Download the current version of an addon locally.
* Display the addon.xml for a remote addon.
* Search all addon.xml files in a repository for a given string.
* Supports the official XBMC respositories as well as 3rd party repos.


Installation
------------

    $ pip install xam


Example Usage
-------------

### List addon ids and versions for the Eden repo

    $ xam all
    * Updating addons.xml from remote...
    metadata.7176.com 1.0.9
    metadata.albums.1ting.com 1.0.7
    metadata.albums.allmusic.com 2.0.1
    ...
    webinterface.wtouch 0.4


### List addon id and version for every addon in bluecop's repo.

    $ xam --repo http://bluecop-xbmc-repo.googlecode.com/files/repository.bluecop.xbmc-plugins.zip all
    * Downloading http://bluecop-xbmc-repo.googlecode.com/files/repository.bluecop.xbmc-plugins.zip to /Users/jbeluch/.xam_cache/aHR0cDovL2JsdWVjb3AteGJtYy1yZXBvLmdvb2dsZWNvZGUuY29tL2ZpbGVzL3JlcG9zaXRvcnkuYmx1ZWNvcC54Ym1jLXBsdWdpbnMuemlw
    * Warning: Repositories which do not zip addons are unsupported at this time. The download functionality might not work properly.
    * Updating addons.xml from remote...
    plugin.video.amazon 0.4.6
    plugin.video.epix 0.1.0
    plugin.video.espn3 0.9.9
    plugin.video.free.cable 0.3.0
    plugin.video.hulu 3.4.5
    plugin.video.nba 0.0.3
    plugin.video.reddit.bc 0.0.3
    plugin.video.vevo 0.7.0
    plugin.video.yahoo.music.videos 0.0.3
    repository.bluecop.xbmc-plugins 1.0.1
    script.hululibraryautoupdate 1.1.2
    script.module.cryptopy 1.2.6
    script.module.demjson 1.4
    script.module.mechanize 0.2.5
    script.module.pycrypto 2.5


### Get the current version for the youtube plugin in the Dharma repo

    $ xam --repo dharma all | grep plugin.video.youtube
    plugin.video.youtube 2.1.4


### List all addons that require [xbmcswift][]
[xbmcswift]: https://github.com/jbeluch/xbmcswift 

    $ xam depends script.module.xbmcswift
    * Local addons.xml is up to date...
    plugin.audio.radioma 1.0
    plugin.video.myvideo_de 0.1.4
    plugin.video.academicearth 1.2.1
    plugin.video.khanacademy 1.4.2
    plugin.audio.radio_de 1.0.6
    plugin.video.aljazeera 0.9.0
    plugin.video.collegehumor 1.0.2
    plugin.video.4players 1.1.3
    plugin.video.eyetv.parser 2.1.1
    plugin.video.nasa 1.0.1
    plugin.video.wimp 1.0.0


### Search for any facebook related addons

    $ xam search facebook
    * Local addons.xml is up to date...
    script.web.viewer:1: Web viewer also allows addon developers to process application authorization (ie. facebook,flickr etc.) with little programming and without violating terms of use.
    script.facebook.media:1: <addon id="script.facebook.media" name="Facebook Media" provider-name="Rick Phillips (ruuk)" version="0.6.4">
    script.facebook.media:2:     <summary lang="en">Browse Facebook photos and videos</summary>


### Download a local copy of the youtube plugin

    $ xam get plugin.video.youtube
    * Local addons.xml is up to date...
    * Downloading http://mirrors.xbmc.org/addons/eden/plugin.video.youtube/fanart.jpg to plugin.video.youtube/fanart.jpg
    * Downloading http://mirrors.xbmc.org/addons/eden/plugin.video.youtube/changelog-2.9.1.txt to plugin.video.youtube/changelog-2.9.1.txt
    * Downloading http://mirrors.xbmc.org/addons/eden/plugin.video.youtube/plugin.video.youtube-2.9.1.zip to plugin.video.youtube/plugin.video.youtube-2.9.1.zip
    * Downloading http://mirrors.xbmc.org/addons/eden/plugin.video.youtube/icon.png to plugin.video.youtube/icon.png
    $ ls plugin.video.youtube/
    changelog-2.9.1.txt
    icon.png
    plugin.video.youtube-2.9.1.zip


### Show the addon.xml for the academic earth plugin
    
    $ xam info plugin.video.academicearth
    * Updating addons.xml from remote...
    Academic Earth (plugin.video.academicearth 1.2.1)
    -------------------------------------------------

    <addon id="plugin.video.academicearth" name="Academic Earth" provider-name="Jonathan Beluch (jbel)" version="1.2.1">
      <requires>
        <import addon="xbmc.python" version="2.0" />
        <import addon="script.module.beautifulsoup" version="3.0.8" />
        <import addon="script.module.xbmcswift" version="0.2.0" />
        <import addon="plugin.video.youtube" version="2.9.1" />
      </requires>
      <extension library="addon.py" point="xbmc.python.pluginsource">
        <provides>video</provides>
      </extension>
      <extension point="xbmc.addon.metadata">
        <platform>all</platform>
        <summary>Watch lectures from Academic Earth (http://academicearth.org)</summary>
        <description>Browse online courses and lectures from the world's top scholars.</description>
      </extension>
    </addon>

### Release a new version of your addon

    (xbmc-vimcasts)jon@lenovo ~/Code/xbmc-vimcasts (master) $ xam release

      oooo    ooo  .oooo.   ooo. .oo.  .oo.
       `88b..8P'  `P  )88b  `888P"Y88bP"Y88b
         Y888'     .oP"888   888   888   888
       .o8"'88b   d8(  888   888   888   888
      o88'   888o `Y888""8o o888o o888o o888o

               (XBMC ADDON MANAGER)

      I'm going to help you create a new release for VimCasts.

    .. Init ..

    Please answer the following:
    [?] I see you are using git. Is master the branch you want to use for the release? (Y/n) 
    [?] This release will be for XBMC EDEN. Is this correct? (Y/n) 

    .. Release Metadata for plugin.video.vimcasts ..

    [?] The current version is 1.1. What should the new version be? [1.2] 1.1
    Writing new version to addon.xml...OK
    [?] I see your addon has a few dependencies. Would you like to check for new versions? (Y/n) 
    Starting new HTTP connection (1): mirrors.xbmc.org
    Dependency script.module.xbmcswift2 is already at the newest version.
    [?] I see you have a changelog.txt. Would you like to update it now? (Y/n) n
    [?] I'm ready to commit the changes and tag the release. Should we continue? (Y/n) y

    .. Release 1.1..

    Adding local changes to staging...OK
    [master d5e32ee] [xam-release-script] creating release for version 1.1
     1 file changed, 6 insertions(+), 7 deletions(-)
    Creating commit...OK
    Tagging release...OK
    Counting objects: 5, done.
    Delta compression using up to 2 threads.
    Compressing objects: 100% (3/3), done.
    Writing objects: 100% (3/3), 385 bytes, done.
    Total 3 (delta 2), reused 0 (delta 0)
    To git@github.com:jbeluch/xbmc-vimcasts.git
       63e631c..d5e32ee  HEAD -> master
    Counting objects: 1, done.
    Writing objects: 100% (1/1), 164 bytes, done.
    Total 1 (delta 0), reused 0 (delta 0)
    To git@github.com:jbeluch/xbmc-vimcasts.git
     * [new tag]         1.1 -> 1.1
    Pushing commit and tags to remote...OK

    Congrats. Release was successful.

    +--------------------+
    | Mailing List Email |
    +--------------------+

    To: xbmc-addons@lists.sourceforge.net
    Subject: [git pull] plugin.video.vimcasts

    *addon - plugin.video.vimcasts
    *version - 1.1
    *url - git://github.com/jbeluch/xbmc-vimcasts.git
    *tag - 1.1
    *xbmc version - eden


CHANGES
Version 0.2.1
-----------

* Add support for language tags.
* Update xam release script


Version 0.2
-------------

* Add `xam release`
* Fix broken imports


Version 0.1
-----------

* Initial release


xam/cli/release.py
'''
    xam.release
    -----------

    Contains the release script to automate the release process for an XBMC
    addon.

    :copyright: (c) 2012 Jonathan Beluch
    :license: BSD, see LICENSE for more details.
'''
import re
import os
import sys
from subprocess import check_call, check_output
import logging

from cliff.command import Command
from colorama import init
init(strip=not sys.stdout.isatty()) # strip colors if stdout is redirected
from termcolor import cprint, colored

from xam.addon import Addon
from xam.repository import get_repo
from xam.cli import REPO_NAMES, add_repo_arg




BLUE = lambda text: colored(text, 'blue')
GREEN = lambda text: colored(text, 'green')
RED = lambda text: colored(text, 'red')

def bump_minor(version_str):
    '''Given a version string, increments the right-most number by 1 and
    returns the new value.

    >>> bump_minor('1.2.3')
    '1.2.4'
    '''
    left, right = version_str.rsplit('.', 1)
    right = int(right) + 1
    return '%s.%s' % (left, right)


def compare_versions(version_a, version_b):
    '''Compares two versions strings and returns 1 if version_a is greater, -1
    is version_b is greater or 0 if they are equal.

    >>> compare_versions('1.0', '1.1')
    -1
    >>> compare_versions('1.1', '0.4')
    1
    >>> compare_versions('0.0.1', '.1')
    0
    '''
    if version_a == version_b:
        return 0

    # Convert parts to ints and strip out blank parts:
    # '.1'.split('.') will result in ['', '1']
    aparts = [int(part) for part in version_a.split('.') if part]
    bparts = [int(part) for part in version_b.split('.') if part]

    # if part lengths are unequal prepend with zeros
    if len(aparts) > len(bparts):
        bparts = [0] * (len(aparts) - len(bparts)) + bparts
    elif len(bparts) > len(aparts):
        aparts = [0] * (len(bparts) - len(aparts)) + aparts

    # same number of parts now, start comparing from leftmost part
    for a, b in zip(aparts, bparts):
        if a > b:
            return 1
        elif b > a:
            return -1
    # the only way we can get here, is if the versions strings aren't equal,
    # but when prepended with zeros they are equal. e.g.: .1 and 0.0.1
    return 0


def write_file(path, contents):
    '''Writes the given contents to the given path'''
    with open(path, 'w') as out:
        out.write(contents)


def generate_email_output(addon_id, version, git_url, tag, xbmc_version):
    '''Prints the message body for a mailing list email.'''
    lines = [
        '',
        '+--------------------+',
        '| Mailing List Email |',
        '+--------------------+',
        '',
        'To: xbmc-addons@lists.sourceforge.net',
        'Subject: [git pull] %s' % addon_id,
        '',
        '*addon - %s' % addon_id,
        '*version - %s' % version,
        '*url - %s' % git_url,
        '*tag - %s' % tag,
        '*xbmc version - %s' % xbmc_version,
        '',
        '',
    ]
    return lines


class ReleaseAddon(Command):
    '''Release a new verison of an addon'''

    log = logging.getLogger(__name__)

    def exit_if_negative(self, msg):
        if not self.yes_no(msg):
            raise RuntimeError('Aborting.')

    def yes_no(self, msg):
        '''Returns False if user entered n or N, True otherwise'''
        ans = raw_input(msg + ' (Y/n) ')
        return ans not in ['N', 'n']
        

    def prompt(self, msg, default=None):
        '''Prints a message and gathers input from the user. The provided default
        is returned if the user provides no input.
        '''
        if default is not None:
            msg = '%s [%s] ' % (msg, default)
        else:
            msg += ': '
        ans = raw_input(msg)
        if ans is not None and len(ans) > 0:
            return ans
        return default

    def get_parser(self, prog_name):
        parser = super(ReleaseAddon, self).get_parser(prog_name)
        add_repo_arg(parser)
        return parser

    def get_cwd_addon(self):
        '''Returns a xam.Addon object for the current addon. Aborts the script
        if no addon.xml file can be found in the current working directory.
        '''
        try:
            return Addon.from_filename('addon.xml')
        except IOError:
            raise RuntimeError('Could not find addon.xml in the current '
                               'directory. Run this script from the root '
                               'directory of your addon.')

    def update_dependencies(self, addon_to_release, xbmc_version):
        '''For any required dependencies, attempts to to update the version number
        to the newest version available in the XBMC official repository.
        '''
        repo = get_repo(xbmc_version)
        addons = repo.addons
        for addon_id, addon_version in addon_to_release.dependencies.items():
            if addon_id != 'xbmc.python':  # skip python dependency
                try:
                    addon = (addon for addon in repo.addons if addon.id == addon_id).next()
                except StopIteration:
                    # ERROR: no addon with that id
                    sys.exit('You have a dependency listed on "%s" but I can\'t '
                             'find an addon with that ID in the %s repository. '
                             'Aborting.' % (addon_id, xbmc_version))

                if compare_versions(addon_version, addon.version) > 0:
                    # ERROR: newest version available is < specified version
                    sys.exit('You require version %s for addon %s, but the newest '
                             'version I can find in the repository is %s. '
                             'Aborting.' % (addon_version, addon_id,
                                            addon.version))
                elif compare_versions(addon_version, addon.version) < 0:
                    # found newer version, prompt for update
                    msg = ('[?] There is a newer version of %s available. '
                           'Would you like to update the dependency from %s to'
                           ' %s?' % (BLUE(addon_id), BLUE(addon_version),
                                     BLUE(addon.version)))
                    if self.yes_no(msg):
                        self.app.stdout.write('Writing to addon.xml...')
                        addon_to_release.set_dependency_version(addon_id, addon.version)
                        write_file('addon.xml', addon_to_release.to_xml_string())
                        self.app.stdout.write(GREEN('OK') + '\n')
                else:
                    msg = 'Dependency %s is already at the newest version.\n' % BLUE(addon_id)
                    self.app.stdout.write(msg)

    def take_action(self, parsed_args):
        '''Performs a release of an XBMC addon.'''
        # Parse addon.xml
        addon = self.get_cwd_addon()
        reponame = parsed_args.repo

        # determine local git branch
        current_branch = check_output('git symbolic-ref HEAD 2>/dev/null',
                                      shell=True)
        if current_branch is None or not current_branch.startswith('refs/heads/'):
            raise RuntimeError('Could not determine current git branch.')
        current_branch_trimmed = current_branch.replace('refs/heads/', '').strip()

        lines = [
            '',
            '''  oooo    ooo  .oooo.   ooo. .oo.  .oo.''',
            '''   `88b..8P'  `P  )88b  `888P"Y88bP"Y88b''',
            '''     Y888'     .oP"888   888   888   888''',
            '''   .o8"'88b   d8(  888   888   888   888''',
            '''  o88'   888o `Y888""8o o888o o888o o888o''',
            '',
            '''           (XBMC ADDON MANAGER)''',
            '',
            '  I\'m going to help you create a new release for %s.' % BLUE(addon.name),
            '',
            '.. Init ..',
            '',
            'Please answer the following:',
            '',
        ]   
        self.app.stdout.write('\n'.join(lines))
        
        # Verify current branch
        msg = '[?] I see you are using git. Is %s the branch you want to use for the release?' % BLUE(current_branch_trimmed)
        self.exit_if_negative(msg)

        # Verify xbmc_version
        msg = '[?] This release will be for XBMC %s. Is this correct?' % BLUE(reponame)
        self.exit_if_negative(msg)

        # Check for dirty repo
        is_dirty = check_output('git status --porcelain', shell=True)
        if is_dirty:
            msg = ('[?] Your git working directory has uncommitted changes. '
                   'These changes will be included in the release commit. '
                   'Would you like to continue anyway?')
            self.exit_if_negative(msg)

        # TODO
        '  * You are using a `git remote` named "origin".',

        # Now onto the release setup
        lines = [
            '',
            '.. Release Metadata for %s ..' % BLUE(addon.id),
            '',
        ]
        self.app.stdout.write('\n'.join(lines) + '\n')

        # version
        msg = '[?] The current version is %s. What should the new version be?' % BLUE(addon.version)
        new_version = self.prompt(msg, default=bump_minor(addon.version))
        addon.version = new_version
        self.app.stdout.write('Writing new version to addon.xml...')
        write_file('addon.xml', addon.to_xml_string())
        self.app.stdout.write(GREEN('OK') + '\n')
    
        # Check dependencies
        if len(addon.dependencies) > 1:
            msg = '[?] I see your addon has a few dependencies. Would you like to check for new versions?'
            if self.yes_no(msg):
                self.update_dependencies(addon, reponame)

        if os.path.exists('changelog.txt'):
            msg = '[?] I see you have a %s. Would you like to update it now?' % BLUE('changelog.txt')
            if self.yes_no(msg):
                check_call([os.getenv('EDITOR'), 'changelog.txt'])

        # Final ok
        msg = '[?] I\'m ready to commit the changes and tag the release. Should we continue?'
        self.exit_if_negative(msg)

        lines = [
            '',
            '.. Release %s..' % BLUE(addon.version),
            '',
        ]
        self.app.stdout.write('\n'.join(lines) + '\n')
    
        # git add -A .
        self.app.stdout.write('Adding local changes to staging...')
        check_call('git add -A', shell=True)
        self.app.stdout.write(GREEN('OK') + '\n')

        # write commit message
        self.app.stdout.write('Creating commit...')
        check_call('git commit -m "[xam-release-script] creating release for'
                   ' version %s"' % new_version, shell=True)
        self.app.stdout.write(GREEN('OK') + '\n')

        # git tag release
        self.app.stdout.write('Tagging release...')
        check_call('git tag -a "%s" -m "%s v%s"' % (new_version, reponame,
                                                         new_version), shell=True)
        self.app.stdout.write(GREEN('OK') + '\n')

        # git push
        self.app.stdout.write('Pushing commit and tags to remote...')
        check_call('git push && git push --tags', shell=True)
        self.app.stdout.write(GREEN('OK') + '\n')

        msg = GREEN('Congrats. Release was successful.')
        self.app.stdout.write('\n' + msg + '\n')

        # attempt to guess public repo URL if using github...  push_url = '<INSERT_REPO_URL>'
        output = check_output('git remote -v | grep origin | grep push | cut -f1'
                              ' -d" "', shell=True)
        match = re.match(r'origin\tgit@github.com:(.+?\.git)', output)
        push_url = RED('<YOUR REPO URL>')
        if match is not None:
            push_url = 'git://github.com/%s' % match.group(1)

        # print email
        output = generate_email_output(addon.id, addon.version, push_url, addon.version,
                    reponame.lower())
        self.app.stdout.write('\n'.join(output))
